zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -0,0 +1,428 @@
+//! Fully populate the Sapling and Orchard note commitment subtrees for existing blocks in the database.
+
+use std::sync::{mpsc, Arc};
+
+use zebra_chain::{
+    block::Height,
+    orchard, sapling,
+    subtree::{NoteCommitmentSubtree, NoteCommitmentSubtreeIndex},
+};
+
+use crate::service::finalized_state::{
+    disk_format::upgrade::CancelFormatChange, DiskWriteBatch, ZebraDb,
+};
+
+/// Runs disk format upgrade for adding Sapling and Orchard note commitment subtrees to database.
+///
+/// Returns `Ok` if the upgrade completed, and `Err` if it was cancelled.
+#[allow(clippy::unwrap_in_result)]
+pub fn run(
+    initial_tip_height: Height,
+    upgrade_db: &ZebraDb,
+    cancel_receiver: &mpsc::Receiver<CancelFormatChange>,
+) -> Result<(), CancelFormatChange> {
+    let mut subtree_count = 0;
+    let mut prev_tree: Option<_> = None;
+    for (height, tree) in upgrade_db.sapling_tree_by_height_range(..=initial_tip_height) {
+        // Return early if there is a cancel signal.
+        if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
+            return Err(CancelFormatChange);
+        }
+
+        // Empty note commitment trees can't contain subtrees.
+        let Some(end_of_block_subtree_index) = tree.subtree_index() else {
+            prev_tree = Some(tree);
+            continue;
+        };
+
+        // Blocks cannot complete multiple level 16 subtrees,
+        // the subtree index can increase by a maximum of 1 every ~20 blocks.
+        // If this block does complete a subtree, the subtree is either completed by a note before
+        // the final note (so the final note is in the next subtree), or by the final note
+        // (so the final note is the end of this subtree).
+        if end_of_block_subtree_index.0 <= subtree_count && !tree.is_complete_subtree() {
+            prev_tree = Some(tree);
+            continue;
+        }
```

Comment:
Consider moving this into an `else` clause at the end of the subtree creation code, to make the condition less confusing.